ts/nni_manager/common/trial_keeper/keeper.ts

```diff
@@ -0,0 +1,175 @@
+ *
+ *  TrialKeeper has a very similar interface to TrainingSerivceV3.
+ *  In fact the local training service is a thin wrapper over TrialKeeper.
+ **/
+
+import { EventEmitter } from 'events';
+import fs from 'fs/promises';
+import path from 'path';
+
+import { HttpChannelServer } from 'common/command_channel/http';
+import globals from 'common/globals';
+import { GpuScheduler } from 'common/gpu_scheduler';
+import { Logger, getLogger } from 'common/log';
+import { TrialProcess } from './process';
+
+export declare namespace TrialKeeper {
+    export interface TrialOptions {
+        id: string;
+        command: string;
+        codeDirectoryName: string;
+        sequenceId?: number;
+        gpuNumber?: number;
+        gpuRestrictions?: GpuRestrictions;
+    }
+
+    export interface GpuRestrictions {
+        onlyUseIndices?: number[];
+        rejectActive?: boolean;
+        rejectComputeActive?: boolean;
+    }
+}
+
+export class TrialKeeper {
+    private envId: string;
+    private channels: HttpChannelServer;
+    private dirs: Map<string, string> = new Map();
+    private emitter: EventEmitter = new EventEmitter();
+    private gpuScheduler: GpuScheduler | null = null;
+    private log: Logger;
+    private platform: string;
+    private trials: Map<string, TrialProcess> = new Map();
+
+    constructor(environmentId: string, platform: string, enableGpuScheduler: boolean) {
+        this.envId = environmentId;
+        this.platform = platform;
+        this.log = getLogger(`TrialKeeper.${environmentId}`);
+
+        this.channels = new HttpChannelServer(this.envId, `/env/${this.envId}`);
+
+        if (enableGpuScheduler) {
+            this.gpuScheduler = new GpuScheduler();
+        }
+
+        this.channels.onReceive((trialId, command) => {
+            if (command.type === 'request_parameter') {
+                this.emitter.emit('request-parameter', trialId);
+            } else if (command.type === 'metric') {
+                this.emitter.emit('metric', trialId, command.metric);
+            } else {
+                this.log.error('Received bad command:', trialId, command);
+            }
+        });
+    }
+
+    public async start(): Promise<void> {
+        await this.channels.start();
+
+        if (this.gpuScheduler !== null) {
+            await this.gpuScheduler.init();
+        }
+    }
+
+    public async shutdown(): Promise<void> {
+        let promises: Promise<void> = [];
+
+        promises.push(this.channels.shutdown());
+
+        if (this.gpuScheduler !== null) {
+            promises.push(this.gpuScheduler.shutdown());
+        }
+
+        const trials = Array.from(this.trials.values());
+        promises = promises.concat(trials.map(trial => trial.kill()));
+
+        await Promise.all(promises);
+    }
+
+    public registerDirectory(name: string, path: string): void {
+        this.dirs.set(name, path);
+    }
+
+    public async createTrial(options: TrialKeeper.TrialOptions): Promise<string | null> {
```

Comment:
I don't like the term trial here. Looks like we are abusing it.

What it actually does is: find a proper device placement and `createProcess`, and keep a channel to communicate with the process.

Reply:
The process has bound trial id and trial command channel. Do you think they should be renamed as well?
P.S. The trial command channel should not be renamed to "command channel", because this channel only accepts trial commands, while there are other channels accepting tuner commands, environment commands, GPU scheduler commands, etc.

Reply:
My concern is that, in distributed trial, this only creates a worker, which is actually part of a trial.

Reply:
Left a reminder. Will rename when add distributed trial support.